fuse_models/include/fuse_models/unicycle_2d_predict.h

```diff
@@ -100,6 +101,105 @@ inline void predict(
   fuse_core::wrapAngle2D(yaw2);
 }
 
+enum StateIndex : uint8_t
+{
+  X,
+  Y,
+  YAW,
+  V_X,
+  V_Y,
+  V_YAW,
+  A_X,
+  A_Y
+};
+
+/**
+ * @brief Given a state and time delta, predicts a new state
+ * @param[in] position1_x - First X position
+ * @param[in] position1_y - First Y position
+ * @param[in] yaw1 - First orientation
+ * @param[in] vel_linear1_x - First X velocity
+ * @param[in] vel_linear1_y - First Y velocity
+ * @param[in] vel_yaw1 - First yaw velocity
+ * @param[in] acc_linear1_x - First X acceleration
+ * @param[in] acc_linear1_y - First Y acceleration
+ * @param[in] dt - The time delta across which to predict the state
+ * @param[in] position2_x - Second X position
+ * @param[in] position2_y - Second Y position
+ * @param[in] yaw2 - Second orientation
+ * @param[in] vel_linear2_x - Second X velocity
+ * @param[in] vel_linear2_y - Second Y velocity
+ * @param[in] vel_yaw2 - Second yaw velocity
+ * @param[in] acc_linear2_x - Second X acceleration
+ * @param[in] acc_linear2_y - Second Y acceleration
+ * @param[in] jacobians - Jacobians wrt the state
+ */
+// TODO(efernandez) fuse this with the templated predict function so we can use the analytic Jacobian in the cost
```

Comment:
I tried that and it didn't work because `predict` is used inside a cost function and Ceres requires the `jacobian` to be a `T**` double pointer: http://ceres-solver.org/analytical_derivatives.html

Reply:
Right. I guess what I was expecting was `predict()` would be wrapped in a derived ceres::CostFunction. The way Ceres populates jacobians is not directly compatible with the new predict() function anyway. So something vaguely along the lines of:
```
class Unicycle2DStateCostFunction : public ceres::CostFunction
{
public:
  bool Evaluate(
    double const* const* parameters,
    double* residuals,
    double** jacobians) const override
  {
    boost::optional<Eigen::Matrix8d> jacobian_matrix = boost::none;
    if (jacobians)
    {
      jacobian_matrix = Eigen::Matrix8d();
    }

    predict(
      position1,
      yaw1,
      vel_linear1,
      vel_yaw1,
      acc_linear1,
      dt_,
      position_pred,
      yaw_pred,
      vel_linear_pred,
      vel_yaw_pred,
      acc_linear_pred,
      jacobian_matrix);  // Only computes jacobians if Ceres wants them

    // Populate the Ceres-requested jacobian blocks
    // I'm having to think about what goes here too hard for a code review.
    // The predict() function computes State2 as a function of State1.
    // But the cost function computes (conceptually) "error = State2 - State1"
    // I'm sure there is a way to translate the predict() jacobian into a Ceres
    // jacobian. Regardless, Ceres wants the jacobian in blocks, not as one
    // giant matrix, so some copying will be required.
    if (jacobians)
    {
      if (jacobians[0])
      {
        Eigen::Map<fuse_core::MatrixXd>(jacobians[0], 8, 2) = jacobian_matrix->block(0, 0, 8, 2);
        // Maybe???
      }
      if (jacobians[1])
      {
        ...
      }
      ...
    }
  }
```

Reply:
Will do that next week. All other changes requested are in. I'm going to validate things still work as intended.